### ISSUE TITLE
DisplayMagician sets the wrong display as the primary display

### ISSUE THREAD

**Issue by domenic** [terrymacdonald/DisplayMagician#23]
**Describe the bug**

DisplayMagician saves a profile which, when restored, always sets the wrong display as primary.

**To Reproduce**
Steps to reproduce the behaviour:

1. Have three displays.
2. Set the right display as disabled, using Windows settings.
2. Set the middle display as primary, using Windows settings.
3. Double-check with NVIDIA Control Panel that the middle display is primary there too.
4. Open DisplayMagician and click "View Current Display"
5. Notice that the diagram shows the *left* display as primary (blue with "[Primary]" text), contradicting the Windows settings/NVIDIA Control Panel.
6. Try saving the profile anyway just to see if it works.
7. Switch to another profile using DisplayMagician.
8. Switch to your saved profile from step 6.

**Expected behaviour**

Unfortunately, it does not work: DisplayMagician sets the left display to primary, instead of the desired middle.

Now I have to go to Windows settings or NVIDIA Control Panel to make the middle display primary after every use of DisplayMagician.

I've tried deleting and recreating the profile with no results.

**Screenshots**

![image](https://user-images.githubusercontent.com/617481/123814632-d7762a80-d8c3-11eb-90bd-dfb078807a52.png)

**Reporting Logs**

[DisplayMagician.log](https://github.com/terrymacdonald/DisplayMagician/files/6734056/DisplayMagician.log)

**Enviroment (please complete the following information):**
 - Windows Version: Win10
 - DisplayMagician Version 1.1.4
 - Video Card: Asus NVIDIA RTX 3080
 - Number and make/model of monitors: 3x Asus ROG SWIFT PG27UQ
 - Date of last video driver update: 2021-06-22
 - Date of last windows update: 2021-06-21

**Additional context**

If I had to guess this might have to do with wanting to set the "higher number" display (3) as primary? In my setup, just because of how I plugged things into the video card, left = 2, middle = 3, right = 1.

**Comment by terrymacdonald** [comment]
Hi Domenic,

DisplayMagician (and HeliosDisplayManagement that DM is based on) uses the
display positioned at position 0,0 (origin) as the indicator of the primary
display. This is because when Windows Display Driver Management changes the
position of the displays, the top left corner of the primary display is
made position 0,0. The position of all the other displays are relative to
that primary display.

But - and it's a big but - as far as I know the primary display isn't
actually set within the code. This part of the codebase isn't something
that DisplayMagician handles itself. The code that actually does the
display setup is actually either NvAPIWrapper or WindowsDisplayAPI
libraries, both of which are written by Soroush Falahati. DisplayMagician
just uses the logic there, and I don't recall exactly what happens; I'd
need to go do some exploring to be sure.

I won't bother with further investigation of this problem with v1.1.4
though as I am actually completely rewriting the display detection/setting
logic as part of the next version of DisplayMagician (v2.0.0). I'm adding
AMD support and HDR support, and as part of that I'm rewriting a large part
of the display management logic. This is unfortunately a huge bit of work
and will take multiple months for me to get to an alpha version that I'll
be able to release for testing. It's also very likely to have multiple bugs
that need ironing out before it is fit for human consumption.

I will add your example scenario to the testing programme for
DisplayMagician v2.0.0, and I'll make sure that the primary display is
stored and applied during profile generation and profile application in
that version.

Please be patient as it will take a few months before I'll be able to build
a solution.

Thanks

*Terry*

**Comment by domenic** [comment]
Sounds good! I'll see if plugging my monitors into different ports works in the meantime...

**Comment by terrymacdonald** [comment]
So some good news here I thought you might like to hear @domenic . Last night I managed to develop some test code in a different test project to test out a new Windows CCD library I put together. The new code does set the display layout and primary monitor correctly, and it also attempts to set the HDR status for each display.

I ran out of time to package it up last night, but if I manage to get that test software ready to go, would you be open to the idea of testing it for me? Do you have an HDR capable screen you can test with? 

The program is called CCDInfo, and when I finish packaging a release zip file it will be here: https://github.com/terrymacdonald/CCDInfo/releases

The idea would be that you would step through the following steps:

- set up a display configuration with HDR enabled on at least one display, and one screen as the primary.
- then you would run 'CCDInfo save <firstfilename>' to generate a config file that would capture the display config, (e.g. `CCDInfo save config1.cfg`)
- then you would change your display configuration to turn off HDR on that one display, and to set a different screen as the primary monito/main display.
- then you would run 'CCDInfo save <secondfilename>' to generate a different config file that would capture the display config as it is (e.g. `CCDInfo save config2.cfg`)

Then the idea is that you would test the ability of CCDInfo to change between those two settings by running the following commands after each other:

- Run `CCDInfo` by itself to list the output, and check that the HDR is on for the one display/
- Change back to the original display configuration using `CCDInfo load config1.cfg`
- Check the output of CCDInfo to check that the HDR state is now off for all displays.
- Open up Windows Display Settings to confirm the settings have changed back. (You will need to close it completely after you make a change, as it doesn't detect the screen change otherwise and things look the same)
- Change back to the different display configuration using `CCDInfo load config2.cfg`
- Check the output of CCDInfo to check that the HDR state is now back on off that single display.
- Close, then reopen Windows Display Settings to confirm the settings have changed back. (You will need to close it completely after you make a change, as it doesn't detect the screen change otherwise and things look the same)

Is that something you can do for me? It will help me confirm the library is working before I then start to build it into DisplayMagician v2.0.0.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@dominec CCDInfo is now released. Please download the release, extract the zip and then run the test that I described in the post above. And then please paste the screen grab of the output too so I can read the bits I need to.

https://github.com/terrymacdonald/CCDInfo/releases/tag/v1.0.0

And thanks for your help!

**Comment by domenic** [comment]
This sounds great! I'm away from my computer for the holiday weekend, but am excited to try testing on Monday or so.

**Comment by terrymacdonald** [comment]
@domenic I've uploaded a working version of CCDInfo (v1.0.1) and have tested it myself. I managed to borrow an HDR monitor and I have validated that CCDInfo will move screen, change the primary screen, change the Windows HDR settings etc. I've tested that it works across reboots and it seems to ok. 

This little programme should be able to fix your problem temporarily while I rewrite DisplayMagician. You can setup a Desktop icon for each of the Display Configs you want (`CCDInfo load nameofconfig.cfg`) and then run them that way, at least until I get DisplayMagician v2.0.0 completed.

The good thing is that I've now fully tested the logic and library functionality that I needed to, so it should make it far easier to update DsplayMagician now!

**NOTE: CCDInfo only sets the Windows HDR settings, so won't affect the NVIDIA HDR settings or AMD HDR settings. Those will be handled by their own libraries that I'll be writing soon.**

**Comment by terrymacdonald** [comment]
@domenic Thanks to some help from Mobeeuz I've discovered that there is a problem applying a confgi once the computer is rebooted. I've created a new version of CCDInfo v.1.0.2 which hopefully solves the problem. Could you please create a couple of configurations and make sure that they both work after a reboot?

Thanks
Terry

**Comment by terrymacdonald** [comment]
@domenic Can you please test CCDInfo v1.0.3 now instead? https://github.com/terrymacdonald/CCDInfo/releases/latest

It should work fine now, but please also reboot and test that it works, and then reboot a 2nd time just to make sure it works after another reboot. I have a suspicion it may not work after that 2nd reboot.

**Comment by domenic** [comment]
I've confirmed that CCDInfo v1.0.4 works great for setting HDR-ness and primary display-ness, by creating saved config files for both variations using `CCDInfo save` and switching between them using `CCDInfo load`.

I was unsure what test you wanted me to run after two reboots. Should I use the old config files and test that loading them still works? Or something else?

**Comment by domenic** [comment]
After two reboots I tried to use `CCDInfo load` and got the following error:

```
CCDInfo v1.0.4
==============
By Terry MacDonald 2021

ProfileRepository/LoadProfiles: Checking whether the display configuration is already being used.
ProfileRepository/LoadProfiles: The requested display configuration is different to the one in use at present. We need to change to the new one.
ProfileRepository/LoadProfiles: Testing whether the display configuration is valid (allowing tweaks).
ERROR - SetDisplayConfig returned WIN32STATUS ERROR_INVALID_PARAMETER while testing that the Display Configuration is valid

Unhandled Exception: System.ComponentModel.Win32Exception: The parameter is incorrect
   at CCDInfo.Program.loadFromFile(String filename) in H:\vscode-projects\CCDInfo\CCDInfo\CCDInfo\Program.cs:line 483
   at CCDInfo.Program.Main(String[] args) in H:\vscode-projects\CCDInfo\CCDInfo\CCDInfo\Program.cs:line 87
```

**Comment by terrymacdonald** [comment]
Please run the exact same Configs that worked before the reboot, but run
them after the reboot.

When the PC reboots, the video card gets assigned a new LUID, and that
change messes up the configuration. I am now finding the new LUID then
patching the adapter LUID in the configuration to make it valid again
before running it. I'd like to test whether that actually works. It seems
to work on my computer, but doesn't on Morbeeus's computer. So I need some
help from you to break the deadlock 😁.

Thanks again for testing for me.

Terry

**Comment by terrymacdonald** [comment]
Darn it. That's the same error Morbeeuz got. Looks like it's back to the drawing board to see what it doesn't like about the Paths or Modes. I'll do more testing until I can get something more reliable. 

The thing I find most weird about this is the fact that it works sometimes and doesn't others!

**Comment by terrymacdonald** [comment]
\o/ I've done it! 

CCDInfo v1.0.5 changes the primary display properly, as well as display layout, resolution, and HDR color settings. Hopefully this is an acceptable workaround until DisplayMagician v2.0.0 is completed.

You can download CCDInfo from here: https://github.com/terrymacdonald/CCDInfo/releases/latest

Some caveats on all the above though; this will only work for Windows Display Settings. CCDInfo will not save or load NVIDIA or AMD settings at all. Those will be coming in some other additional test programs later on as I carry on with the AMD development (AMDInfo first, then NVIDIAInfo next). Once those are tested fully, then they'll be integrated into DisplayMagician v2.0.0.

Now that I have competed the first library for DisplayMagician (WinLibrary), and I'll be back working on the AMD one shortly (AMDLibrary).

**Comment by terrymacdonald** [comment]
Please test CCDInfo 1.0.5 @domenic and if this is now working, please close this issue. I am working to add primary display support to DisplayMagician as part of v2.0.0 and that will be released as part of the normal release cycle. Current ETA is just under 2 months at a guess.

Terry

**Comment by terrymacdonald** [comment]
@domenic  I've just finished the first 'basic' version of NVIDIAInfo v0.0.1 Alpha. Can you please test it for me to see whether it works with your NVIDIA Surround setup? I've written a new NVIDIA library from scratch, and this is the first 'mostly' working version. It will successfully save and load NVIDIA surround/mosaic settings, but it won't do NVIDIA HDR or anything else. It is a pre-release version of the software, but I need you to do some testing so I can be sure its working.

Please download the NVIDIAInfo v0.0.1 zip file from https://github.com/terrymacdonald/NVIDIAInfo/releases/tag/v0.0.1, unzip the zip file and then follow the instructions in the release notes to save 2 different display profiles - one using surround, and one not using surround. Then please try to swap between them using the load command.

Once you've finished testing, please post your NVIDIAInfo.log files here for me to see, and tell me how the testing went.

Thanks
Terry

**Comment by domenic** [comment]
The surround stuff works, but when I switch back to three-monitor-mode it keeps resetting my primary screen to the wrong screen, which I then need to fix using Windows display settings.

[NVIDIAInfo.log](https://github.com/terrymacdonald/DisplayMagician/files/6994508/NVIDIAInfo.log)

**Comment by terrymacdonald** [comment]
I *think* that's already been fixed in the codebase, but I'll need to
check. It will be a few days before I'll be able to work on the other bits
I need to complete before the next release. Hopefully I'll have a new
working version for you to test in a week or so.

**Comment by terrymacdonald** [comment]
@domenic  I've added the ability to save NVIDIA HDR information to NVIDIAInfo v1.0.0, as well as checking that the display configuration is available before attempting to apply it. I've tested it successfully internally, but now I need you to test it again for me to make sure I haven't affected the GeForce card support. I have successfully tested that it changes the primary display properly when changing screens on my test machine, but now I need you to test it to be sure.

**IMPORTANT - The configuration file format has changed so you will need to create new cfg files for this as the old configs will not work with NVIDIA v1.0.0.**

So, can you please download `NVIDIA v1.0.0.zip` from https://github.com/terrymacdonald/NVIDIAInfo/releases/tag/v1.0.0, and then test that it works?

Please create four new configuration files for the following scenarios:

- A config file for a Mosaic screen with all 3 screens joined as a single screen
- A config file for a Triple screen with the left screen as the main screen
- A config file for a Triple screen with the middle screen as the main screen
- A config file for only two screens with the right screen turned off

Please test swapping between those 4 different configs to make sure that it works correctly transitioning between each configuration.

If it all works without any issues, then I guess I'm moving on to integrating this library with DisplayMagician!

**Comment by domenic** [comment]
It's almost perfect; probably at least as good as what's in DisplayMagician right now. I'm considering switching from DisplayMagician at this point, except for one of my game-specific shortcuts, because it gets the primary screen thing right.

The only scenario I had trouble with was when I set up my displays as mosaic, with HDR set to on in the Windows display settings, and bit depth set to 10 bits in the NVIDIA control panel. When I tried to load that config, I would get

```
ERROR - NVIDIAInfo Windows CCD settings were not applied correctly.
```

However, checking the Windows display settings and the NVIDIA control panel showed everything seemed to be correct. So, I'm not sure what happened there.

Logs from that run are attached. (I guess you overwrite the log file each run?)

[NVIDIAInfo.log](https://github.com/terrymacdonald/DisplayMagician/files/7034516/NVIDIAInfo.log)

**Comment by terrymacdonald** [comment]
Hi @domenic,

Thanks for testing. Seems like I was so close to getting HDR right, but so far away :). I did have an idea that there may be some sort of conflict between them both if I added HDR, but I thought I'd give it a go anyway. Never hurts to try.

Thanks for the NVIDIA.log, it definitely helps. I have realised that I need your surround-10bpp.cfg so I can see what settings it contains. That file will help me troubleshoot the HDR error better, and hopefully I'll be able to get it working. The HDR error appears to be the last problem with the NVIDIA library, so I'm hopeful I'm getting closer to a DisplayMagician release.

On that note, I had my first big win, as I managed to get the existing NVIDIA library integrated into DisplayMagician (ignore the AMD symbol at the top right!) :

![image](https://user-images.githubusercontent.com/5721772/130605401-911aaa04-7f86-4943-a75f-2c2bd211b836.png)

I think it looks good! There are still a LOT of errors I need to fix in order to get things working. My gut feel is it will be another month before I'll be able to get DisplayMagician v2.0.0 released, but as you can see I'm making pretty good progress. You'll soon get one program that does it properly.

Thanks
Terry

**Comment by terrymacdonald** [comment]
And another update. This has been a monster challenge, but I think I'm nearing the end of the NVIDIA chapter at least. Things that the new DisplayMagician can do:

Swap between surround and non-surround profiles within about 14 seconds
Handle surround and non-surround screens together
Works with NVIDIA mosaic cards now
Uses the PCI video card vendor ID to recognise NVIDIA and AMD vendors, and sets it's driver mode accordingly
Can be 'forced' to use a different video card mode if needed (unlikely to be used, but hey, why not)
Can set/clear a Desktop Background per profile if you want
Stores and applies Windows HDR and NVIDIA HDR settings
Records all settings provided by the NVIDIA driver and windows drivers under Windows 10, meaning that you display refresh settings are stored and set, as are bezel distances, or any other weird and wonderful settings you creative users are able to configure :)
But at present there are still bits I have to finish before I can craft any sort of production ready release:

It is very slow to load. I've identified this is due to DisplayMagician keeping on calling the video card libraries. I need to implement some caching to ensure that the video detection is only run at programme launch and when screen layout changes in some way.
I need to add AMD library support. This is the core reason why I started this journey, so it seems a little disingenuous to release DisplayMagician without this!
I need to add a nice message that popups when an old DisplayMagician file is detected, that informs the user that this is a new version of DisplayMagician that uses a brand new DisplayProfile format that will require them to create a brand new set of Display Profile layouts.
I need to test, and to provide my beta testers an early release of the software so that I iron out any bugs before a general release.
I need to add the ability to test the upgrade mechanism without upgrading the production upgrade functionality. This is needed so I can iron out any issues before a general release. Very important as DisplayMagician now has over 1500 users!
And I think that's about it!

Some cool things to come for sure, and it's taken a lot of hard work to get this close. Another little while and we're hopefully good to go!

Thanks
Terry

**Comment by terrymacdonald** [comment]
@domenic,

I finally have an test version of DisplayMagician v2.0.0 to share, and I'd be honoured if you'd do some testing for me.

I've attached DisplayMagician v2.0.0-alpha1.msi to this issue, and I'm really interesting to find out if it works on your system. This version of DisplayMagician uses two completely new display manipulation libraries that I created (one for NVIDIA and one for Windows CCD) so it needs some proper testing. Please feel free to test complicated setups like 2 screens in a Surround screen with another two screens next to it, or with different refresh rates, or HDR on only some of the screens. **I really want to know if it will break.**

IMPORTANT: DisplayMagician v2.0.0 uses a completely new Display Profile format. This means that you will need to create brand new Display Profiles when you first start DisplayMagician v2.0.0, and you will then need to edit your Game Shortcuts to use those new Display Profiles.

**How to install this version of DisplayMagician:**

- Download and run the DisplayMagicianSetup-v2.0.0-alpha1.msi installer
- Once installed, run DisplayMagician v2.0.0
- Go straight to 'Settings' on the main screen
- Change the Logging Level to 'Full application trace'
- Click 'Back'
- Click the 'Display Profiles' button
- Setup the Display Profiles that you want to use (e.g. open up Windows Setup or NVIDIA Setup and change your display configuration until it's the way you want it, then swap back to DisplayMagician and save the DisplayProfile).
- Edit your Game Shortcuts to use the new Display Profiles you just set up.

Then now please start using DisplayMagician v2.0.0 and see what breaks!

If you find a bug, please log it as a new issue at https://github.com/terrymacdonald/DisplayMagician/issues/new/choose

Thanks
Terry

[DisplayMagicianSetup-v2.0.0-alpha1.zip](https://github.com/terrymacdonald/DisplayMagician/files/7139882/DisplayMagicianSetup-v2.0.0-alpha1.zip)

**Comment by terrymacdonald** [comment]
One thing I forgot to mention is that it currently won't work on any system that has an AMD video card. I've not finished the AMD library, yet AMD detection is already plumbed into DisplayMagician, which means that DisplayMagician will attempt to use the non-existent AMD library if an AMD video card is detected, and it will just crash.

Thanks
Terry

**Comment by domenic** [comment]
I installed the alpha and it generally works great! I created profiles for my three setups (3 separate monitors, 1 surround, 2 separate monitors with the right one off for work) and can switch between them using the new desktop shortcuts with no problem.

I also updated my surround game profile to use the new surround display profile and it seems to work OK, although I noticed some delays. I will continue using it going forward and report back if I can come up with any concrete reproducible issue.

Overall this is a big improvement: the primary display is set correctly (fixing the original issue this thread is about!) and all switches are much much faster. I'm not sure if there were any HDR issues before in v1, but knowing that HDR works fully now is great too. I love it!

**Comment by terrymacdonald** [comment]
@domenic Great news! DisplayMagician v2.0.1 has been released, and fixes this very issue. Please test it out, and if you still have any issues please log a new issue.

I'm closing this issue as the fix is in DisplayMagician v2.0.1, and I need to clear out the completed issues so I am able to work on the new issues that come in.

Thanks
Terry
